Feature/highlight and capture

joshuanoeldeke opened the pull request.

joshuanoeldeke committed:
feat: Add configurable live camera highlighting

Implements a visual highlight (green border) on the live camera view when a recognized object matches user-defined criteria.

- Adds a "Highlight Rules" section in the Settings tab where users can configure which object labels and confidence thresholds trigger the highlight.

- The settings UI includes label validation and autocomplete against the active ML model's known labels.

- Users can add, delete, and edit the highlight rules.

- Includes several refactors and bug fixes in the model loading services that were discovered during implementation.

joshuanoeldeke committed:
feat: Add model persistence and best shot settings

- Persists the user's selected ML model choice across app launches by centralizing state in AppStateViewModel.

- Adds a new setting for the "Best Shot" feature, allowing the user to specify a single target label for auto-capture.

- Implements a new settings view for the target label with an autocomplete UI that suggests valid labels from the current model.

- Refactors the "Best Shot" capture logic to use this new dedicated target label.

- Implements the basic UI and logic for the "Best Shot" capture sequence, including a countdown timer and a results view.

joshuanoeldeke committed:
feat: Implement high-resolution capture for Best Shot

- Refactors the "Best Shot" capture sequence to use the high-resolution AVCapturePhotoOutput.

- This replaces the previous method of capturing frames from the low-resolution video stream.

- Updates the photo saving mechanism to use PHPhotoLibrary, which preserves all image metadata (e.g., location, date, camera settings).

- Saved images will now appear in the user's photo library with the same quality and detail as if taken by the native Camera app.

- Updates the CaptureCandidate data model to handle raw image Data and modifies the BestShotResultsView accordingly.

joshuanoeldeke committed:
feat: Add Assisted Capture mode

- Implements a new "Assisted Capture" mode.

- When enabled via a new toggle in Settings, the main shutter button is disabled by default.

- The shutter is only enabled when an object matching the user's "Highlight Rules" is detected, providing a visual cue to the user for taking a good photo.

- This feature is tied directly to the existing highlight system for a seamless experience.

joshuanoeldeke committed:
chore: Remove obsolete scripts and configure xcconfig

- Deletes create_viewmodels.sh and create_all_remaining.sh. These scripts were used for initial project setup and are now outdated.

- Links the DeveloperSettings.xcconfig to the project build configurations.

joshuanoeldeke committed:
refactor: Overhaul manual capture to save to Photo Library

Refactors the manual photo capture workflow to align with native camera behavior. Manual captures are now saved directly to the device Photo Library with full metadata. The camera view no longer dismisses after a capture, allowing for continuous shooting. Adds a temporary "Saved" confirmation toast to the UI. Decouples the manual capture flow from the home screen analysis view. Fixes a build error by restoring the processLiveClassifications function.

joshuanoeldeke committed:
docs: Update README and perform code cleanup

- Updates the README.md with the latest feature list and a detailed architecture overview.

- Refactors `CameraViewModel` to `HomeViewModel` to more accurately reflect its role in managing the Home tab.

- Deletes the unused `CameraService.swift` file to remove redundant code.

joshuanoeldeke committed:
feat: Add haptic feedback and enhance UI polish

- Implements a centralized `HapticManager` for providing haptic feedback.

- Adds haptic feedback to key user interactions:

  - Manual photo capture.

  - "Best Shot" sequence start, stop, and candidate capture.

  - Countdown ticks for the last 3 seconds of the "Best Shot" sequence.

  - ML model selection in Settings.

- Enhances the "Best Shot" countdown overlay to display the target label and the number of candidates found, improving user feedback.

joshuanoeldeke committed:
fix: Revert timer logic to stable version and fix performance

- Reverts the "Best Shot" countdown timer to a simpler, more predictable implementation to resolve UI stability issues.

- Fixes a major performance issue by creating thumbnails on a separate background thread, preventing the main thread from freezing.

- Fixes thumbnail distortion by ensuring aspect ratio is maintained.

- Removes the arbitrary 3-photo limit on "Best Shot" results.

- Removes temporary debugging logs.

joshuanoeldeke committed:
feat: Implement Best Shot cancel button and easy cleanup

- Implements a cancel button for the "Best Shot" sequence, allowing users to stop the countdown early.

- The "Best Shot" button now toggles between starting and stopping the sequence, with a visual change in its icon and color.

- Performs easy cleanup by removing the unused `HomeViewModel` dependency from `LiveCameraView`, simplifying its interface.

- Fixes access control for `stopBestShotSequence` to allow UI interaction.

joshuanoeldeke committed:
feat: Embed location metadata in saved photos

- Integrates CoreLocation to capture the device's location.

- Requests "When In Use" location authorization from the user.

- Embeds the captured CLLocation data into photos saved to the Photo Library, ensuring they appear with location metadata in the Photos app.

- Updates CaptureCandidate to store location data.

- Starts and stops location updates with the camera session.

- Adds NSLocationWhenInUseUsageDescription to Info.plist.

joshuanoeldeke committed:
Clean up Services/Utilities

joshuanoeldeke committed:
chore: refactor component names

joshuanoeldeke committed:
refactor: massive code cleanup and modularization

BREAKING CHANGES:
- Renamed swift_camApp to SwiftCamApp (proper Swift naming)

Major improvements:
- Split ContentView.swift from 1,008 lines to 59 lines (94% reduction)
- Extracted HomeTabView, CameraTabView, SettingsTabView into separate files
- Created 6 reusable UI components:
  * PremiumEmptyStateView
  * ScaleButtonStyle
  * ModelSettingRow
  * InfoRow
  * CameraSettingToggleRow
  * BlurStyleRow
- Added comprehensive documentation to all ViewModels
- Fixed all @Previewable warnings (zero warnings in build)

Documentation:
- Added CODEBASE_ANALYSIS.md with architecture deep dive
- Added REFACTORING_SUMMARY.md with before/after metrics
- Added PRESENTATION_GUIDE.md with demo script and Q&A

Build status: ✅ SUCCESS (0 errors, 0 warnings)

joshuanoeldeke committed:
refactor: extract services and fix memory/threading issues

Service extraction:
- Created CameraSessionManager for camera lifecycle management
- Created BestShotService for automatic capture sequences
- Created VisionService to eliminate code duplication
- Split AppStateViewModel into coordinator + UserSettingsViewModel

Memory & threading fixes:
- Fixed potential retain cycle in LiveCameraViewModel Vision request
- Changed nested DispatchQueue.main.async to Task {@MainActor}
- Added explicit 'self' capture in closures

Architecture improvements:
- AppStateViewModel now coordinates initialization + settings
- UserSettingsViewModel handles all persistent preferences
- Services use callback pattern for better separation
- All services fully documented

Build status: ✅ SUCCESS (0 errors, 0 warnings)

joshuanoeldeke committed:
Delete MDs directory

joshuanoeldeke committed:
docs: add comprehensive documentation to all views and components

Complete documentation coverage achieved:

Views Documentation:
- SwiftCamApp.swift: App lifecycle and initialization flow
- ContentView.swift: Tab navigation architecture
- LiveCameraView.swift: Camera features with UI diagram
- HomeTabView.swift: Photo classification flow
- CameraTabView.swift: Tab wrapper pattern
- SettingsTabView.swift: All 6 settings sections
- SplashScreenView.swift: Loading sequence

Components Documentation:
- CaptureButton: iOS-style shutter button
- ZoomControlView: Multi-camera zoom selector
- CameraPreviewView: AVFoundation bridge
- ClassificationRow: Result display with color coding
- ConfidenceBadge: Confidence indicator
- ImagePreviewView: Image display with loading
- HomeClassificationResultsView: Results orchestration
- LiveClassificationResultsView: Real-time results
- ResultsList: Formatted results list
- ErrorView: Error message display
- EmptyResultsView: Empty state placeholder
- SettingsNavigationRow: Reusable nav component

Documentation Quality:
✅ Class-level overviews with purpose
✅ Feature descriptions
✅ Usage examples with code
✅ Visual design notes
✅ Integration patterns
✅ State handling explained

Result: 100% documentation coverage across all layers
Build status: ✅ SUCCESS (0 errors, 0 warnings)

joshuanoeldeke committed:
feat: Fix face blur coordinate system and improve privacy settings

Major Improvements:
- Fix face blur positioning in captured photos by properly handling image orientation
- Separate live preview blur from captured photo blur for better performance
- Add blur style selection UI with visual feedback
- Increase blur strength for better privacy protection (1.5x Gaussian, 2x Pixelation)
- Add toggle for location metadata embedding in settings
- Reorganize settings into logical groups (Camera, Privacy, Best Shot, Advanced)

Technical Changes:
- FaceBlurringService: Use oriented CIImage to match Vision's coordinate space
- LiveCameraViewModel: Add separate livePreviewBlurEnabled and faceBlurringEnabled flags
- UserSettingsViewModel: Add includeLocationMetadata setting with persistence
- BlurStyleRow: Add selection indicator for active blur style
- SettingsTabView: Group settings into sections for better UX
- HomeTabView: Move model status inline with header for cleaner layout

Bug Fixes:
- Face blur now appears in correct position in captured photos
- Blur style selection now shows active state in UI
- Live preview blur no longer freezes camera feed
- Front camera preview orientation corrected

mauriba commented on swift-cam/Models/AppConstants.swift:
```diff
@@ -9,7 +9,7 @@ import CoreGraphics
 enum AppConstants {
     static let preloadModels = true
-    static let maxClassificationResults = 2 // Show only top 2 predictions
+    static let maxClassificationResults = 3 // Show only top 2 predictions
```
... maybe remove the comment, as it is now off by one

mauriba commented on swift-cam/ViewModels/AppStateViewModel.swift:
```diff
@@ -30,37 +109,44 @@ class AppStateViewModel: ObservableObject {
                 
+                // Parse progress (e.g., "Loading MobileNet V2... (1/3)")
                 if let match = progressText.range(of: "\\((\\d+)/(\\d+)\\)", options: .regularExpression) {
                     let numbers = progressText[match].dropFirst().dropLast().split(separator: "/")
```
Why are we parsing the progression of ML model loading from a string?

joshuanoeldeke replied:
Because thats what ModelPreloader returns currently... Working on changing it :)

mauriba commented on swift-cam/Views/Components/CameraPreviewView.swift:
```diff
@@ -2,13 +2,32 @@
 //  swift-cam
 //
-//  Camera preview layer wrapper for AVCaptureSession
+//  UIKit-wrapped camera preview layer for AVCaptureSession
```
This comment is very useful, thanks! :D

mauriba reviewed (approved):
Amazing work!
I really like the new live camera modes and all the doc comments added :D

joshuanoeldeke committed:
refactor: Reorganize views and improve concurrency

- Reorganizes view components into feature-specific folders.
- Refactors the Vision and camera processing pipeline to improve performance and adopt modern Swift concurrency practices.
- Caches VNCoreMLRequest objects in VisionService to prevent expensive re-initialization and stop ANE power cycling.
- Rewrites the LiveCameraViewModel frame processing logic to be fully concurrency-safe, resolving all Sendable and actor-isolation warnings.
- Restores the live preview and face blur features that were accidentally removed during the refactoring.
